### PR TITLE
[Arrow] Add ChunkScanState interface to preserve chunk-offset when scanning

### PR DESCRIPTION
**Pull request by Tishj**
This PR adds an abstract ChunkScanState class, which can be used when scanning from a source that produces data chunks, and the chunks could potentially only be partially processed.

Previously we used a `CurrentChunk` struct located in the QueryResult to keep this state, this was only used in the `ArrowUtil::FetchChunk` method.
We now move to using a `ChunkScanState` in there instead, which allows us to replace the source that these chunks come from.

This is currently only implemented for one source, the `QueryResult`.
Previously we used `CurrentChunk` which was being kept in the QueryResult, this PR retires that struct.

Motivation for this PR is work on the upcoming pyarrow result collector.
In that PR we will need to scan from a ColumnDataCollection in `ArrowUtil::FetchChunk`, and we will need similar state to this.

### CODE REVIEW

**Review by pdet — COMMENTED** [review]
Hi @Tishj thanks for the PR, it looks great! I just had some small comments.

You might also need to update the arrow ipc extension

File: src/main/chunk_scan_state.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	}
+	return result.TryFetch(current_chunk, error);
+}
+
+bool QueryResultChunkScanState::HasError() const {
+	return result.HasError();
+}
+
+PreservedError &QueryResultChunkScanState::GetError() {
+	return result.GetErrorObject();
```
[the remark this answers is not in the document]
Comment (Tishj) [reply src/main/chunk_scan_state.cpp:68]: I think GetErrorObject already does that, I ran into assertions here when working on this

File: src/main/chunk_scan_state.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+vector<LogicalType> &QueryResultChunkScanState::Types() {
+	return result.types;
+}
+
+vector<string> &QueryResultChunkScanState::Names() {
+	return result.names;
+}
```
[the remark this answers is not in the document]
Comment (Tishj) [reply src/main/chunk_scan_state.cpp:77]: Yes you're right, I've adjusted them in the PR where I depend on this, will push it upstream to this branch as well

**Comment by Tishj** [comment]
> Hi @Tishj thanks for the PR, it looks great! I just had some small comments.
> 
> You might also need to update the arrow ipc extension

I had a look at the `duckdblabs/arrow` extension repo and compiled with it, no errors and it doesn't seem to be affected by this PR. It only deals with the ArrowAppender and this abstracts it away from the implementation details

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**